### ISSUE TITLE
_FillValue attribute creation error for NC_GLOBAL with netCDF 4.5.x?

### ISSUE THREAD

**Issue by czender** [Unidata/netcdf-c#458]
netCDF library behavior of _FillValue seems to have changed in a backwardly incompatible way. With the 20170804 netCDF snapshot this command fails to create a global _FillValue attribute in a netCDF3 file, whereas the same command succeeds with netCDF 4.4.x and earlier versions. Is this new behavior intended? 
```
zender@skyglow:~$ ncatted -O -a _FillValue,global,c,l,222 ~/nco/data/in.nc ~/foo.nc
nco_err_exit(): ERROR Short NCO-generated message (usually name of function that triggered error): nco_put_att()
nco_err_exit(): ERROR Error code is -50. Translation into English with nc_strerror(-50) is "NetCDF: Action prohibited on NC_GLOBAL varid"
nco_err_exit(): ERROR NCO will now exit with system call exit(EXIT_FAILURE)
```

**Comment by WardF** [comment]
Thanks I will track this down!

**Comment by czender** [comment]
Thank you. FYI 4.5.x also causes a less clear regression with ncpdq, which now fails to pack variables that have a _FillValue attribute. This is a major regression, since people do this all the time. I will try to narrow down what it happening further when I have time.

**Comment by DennisHeimbigner** [comment]
Do I have this right? If in.nc was an empty netcdf file, the ncatted command
would be equivalent to this?
    ncgen -3 in.cdl
where in.cdl would be:
netcdf test {
// global attributes:
		int64 :_Fillvalue = 222LL ;
}

**Comment by czender** [comment]
Yes, except the _FillValue would be a plain int not int64.

**Comment by czender** [comment]
I now understand that this error is due to #388. FWIW it has always been (or seemed) legal to have a _FillValue defined for a NC_GLOBAL, and NCO supported this. A consequence is that there may be files in the wild with _FillValue defined for root and groups. #388 may make copying/manipulating such files difficult, since an error code will be returned. Does this backwards incompatibility merit revisiting #388?

**Comment by WardF** [comment]
@czender Absolutely worth revisiting and a discussion; the files can still be read correctly?  We need to avoid breaking the backwards compatibility, and depending on the actual harm done we may just revert #388 and instead update the documentation.  @DennisHeimbigner any thoughts? We can discuss this at our meeting if not here.

**Comment by czender** [comment]
I patched the latest NCO snapshot to emit a warning when it receives an NC_EGLOBAL error from nc_put_att(NC_GLOBAL,"_FilllValue"), but otherwise proceed normally. This helps to mitigate the issue.

**Comment by DennisHeimbigner** [comment]
In order to  keep https://github.com/Unidata/netcdf-c/issues/388
and to usably support back-compatibilty, I think we need to
change from reporting an error to reporting a warning somehow.
If the file is being generated with ncgen, then ncgen can report the
warning. If it is via the library, then our only option is to use logging
to report the warning. Unfortunately, logging is netcdf-4 only.
In any case, attempts to write such a global _FillValue could be silently
ignored.
BTW, maybe we should consider extending logging support to all parts
of netcdf library and not just netcdf-4. Otherwise we have not way to
report warnings.

**Comment by WardF** [comment]
I don't think silently ignoring is a good idea; if data isn't written it should be reported somehow, or else we set ourselves up for future problems when people think they are doing one thing but are really doing another.  Maybe we could have a hidden attribute in the file that is used to store warning messages, etc?  There's probably a problem with that, I haven't really thought it through yet.

**Comment by DennisHeimbigner** [comment]
It occurs to me to ask: did a global _FillValue actually get used in defining
the fillvalue for any variable? If so, then I think we have to keep that behavior.
If it was effectively unused, then we can "ignore" in some sense that global
fillvalue.

**Comment by czender** [comment]
I'm not sure if Dennis' question was for me so let me clarify that currently/formerly people could have groups with a _FillValue with no error. Some of them may have workflows that cause their variables to "inherit" the _FillValue of the group they are in. I don't know of any specific examples.

**Comment by DennisHeimbigner** [comment]
I guess I was trying to figure out what behavior needs to be preserved.
If we allow setting a global _FillValue, but otherwise ignore it, does that
preserve back compatibility?

**Comment by czender** [comment]
NCO exposes the full attribute API, i.e., creating, deleting, copying, renaming attributes. If the full attribute API is allowed to continue to work with _FillValue on NC_GLOBAL then nothing changes from the user's point of view, so that would be back-compatible. Eliminating any of those could potentially cause grief to some users.

**Comment by WardF** [comment]
NCO is a vital suite of tools; I'm convinced that we need to find a solution (such as Dennis suggests) that does not break the tools or user workflows which adhere to what has become a de-facto standard.  As far as I can tell, the global `_FillValue` attribute isn't used for anything internally, and as such can be ignored by the library.  My approach for the next rc3/full release would be (absent a reason to do something else): updating the documentation to reflect this and reverting the previous pull request that prevents setting a global `_FillValue` attribute.

**Comment by DennisHeimbigner** [comment]
Sounds right to me.

**Comment by WardF** [comment]
I've created a temporary branch, `reverse-388`.  Does this fix the issue you're seeing re: `ncpdq` @czender ?  I need to do more to understand the underlying issue but I want to make sure I'm looking in the right place to start dissecting the code.

**Comment by czender** [comment]
There is another, though possibly related, issue caused by changed treatment of _FillValue in 4.5.x that is, IMHO, both less straightforward and a bigger deal than the NC_GLOBAL issue discussed here (in #458) and in #388. I'm uncertain precisely what netCDF change caused this (possibly #383 and/or #384 ???). As background, NCO allows the _FillValue type to be other than the variable type, and ncpdq does not try to change the _FillValue type to the packed type. 4.5.x no longer allows this, and now emits NC_EBADTYPE (-45): "NetCDF: Not a valid data type or _FillValue type mismatch" from ncpdq's nc_put_var() that writes the packed data to the variable that already has a _FillValue (of a different type) defined. Many people (including me) rely on this feature, although it is not a best practice (I wrote ncpdq in 2004, when missing_value was still the norm, and possibly before _FillValue best practices were defined). So, I consider this issue is partly due to an NCO design flaw (being lazy and not always making typeof(_FillValue) == typeof(variable)). Still, the current algorithm works pretty well when _FillValue is in the range of the packed type (otherwise ncpdq does print a warning). Modifying ncpdq is possible, but could take months. In the meantime, many workflows with ncpdq will break if/when it's linked to 4.5.x. I'm unsure how to proceed.

**Comment by WardF** [comment]
My first thought off the top of my head is to back out the change implemented re: different types and instead phase it in slowly while flagging the behavior as deprecated, and work together towards not locking this behavior out altogether until NCO has been updated. Although depending on the scope of the problems it could cause, I am hesitant to do something that breaks widespread workflows built around behavior that was permitted, even if it was documented as not a best practice. Even if it's not a best practice, unless it was explicitly forbidden, i dont know how easily we can/should change the behavior at this point. It will merit more discussion once I'm done with this symposium, but I am inclined to balance the damage caused by enforcing this vs. the hypothetical damage prevented. Let me refer back to those other pull requests when I can, to refamiliarize myself with the motivation for making the changes.

**Comment by DennisHeimbigner** [comment]
Can you elaborate on what you mean by the term "packed"?

**Comment by czender** [comment]
scale_factor and/or add_offset used to convert floating point to smaller type.

**Comment by DennisHeimbigner** [comment]
I was afraid of that. The whole scale+offset thing is a nightmare because
the spec did not type the variable to be float or double, which its true type.
At the moment, I have no idea how to address this.

**Comment by WardF** [comment]
I think for the short term we need to revert the pull request which has caused this issue, and then proceed more slowly once we have identified a potential fix/approach.  Clearly this behavior is too widespread to simply change without causing problems.

Good news, the issue is not #383 as that has not been merged in yet.  I will evaluate the changes in #384.  @czender is there an easy way to test against `ncpdq`? Will `make check` in NCO report an error?  I suspect not since NCO is part of our normal regression testing, but perhaps there is something else to check?

**Comment by czender** [comment]
Sounds good. I'll release NCO 4.6.8 as is, and can alter ncpdq in a later version coordinated to work with new behavior in netCDF 4.5.x. 
Both the regressions discussed on this page were identified with the NCO regression tests, which are invoked by "make test" (not "make check", which tests something else). There are a few expected failures in some operators due to known bugs, but ncpdq has no regressions with 4.4.x and only one regression (test 31) with 4.5.x. That's how I identified the issue originally. But that's a pretty high overhead to test this issue and so yesterday I wrote a targeted little script to demonstrate the problem and guide future changes in ncpdq:
```
cat > ~/ncpdq_tst.sh << 'EOF'
# Usage:
# ncpdq_tst.sh var_nm
ncks -H -C -m --cdl -v ${1} ~/nco/data/in.nc
ncpdq -O -C -v ${1} ~/nco/data/in.nc ~/foo.nc
ncks -H -C -m --cdl -v ${1} ~/foo.nc
ncpdq -O -U -C -v ${1} ~/foo.nc ~/foo_upk.nc
ncks -H -C -m --cdl -v ${1} ~/foo_upk.nc
EOF
chmod a+x ~/ncpdq_tst.sh

```
This script prints the CDL from a single variable in the file ~/nco/data/in.nc (easy to change) in three states: 1. Before Packing with ncpdq, 2. After packing with ncpdq, 3. After unpacking with ncpdq. The regression discussed here causes the packing step to fail (noisily). ncpdq always fails to pack a variable with any data equal to _FillValue fails with 4.5.x, and always succeeds with 4.4.x, so perhaps you just test with `ncpdq -v var_nm in.nc out.nc`. With 4.4.x, the script output looks like this:
```
zender@firn:~$ ~/ncpdq_tst.sh fll_val
netcdf in {
  dimensions:
    lon = 4 ;

  variables:
    float fll_val(lon) ;
      fll_val:long_name = "_FillValue example" ;
      fll_val:_FillValue = -999.f ;

  data:
    fll_val = 73, _, 73, _ ; 

} // group /
netcdf foo {
  dimensions:
    lon = 4 ;

  variables:
    short fll_val(lon) ;
      fll_val:long_name = "_FillValue example" ;
      fll_val:_FillValue = -999.f ;
      fll_val:add_offset = 73.f ;

  data:
    fll_val = 0, _, 0, _ ; 

} // group /
netcdf foo_upk {
  dimensions:
    lon = 4 ;

  variables:
    float fll_val(lon) ;
      fll_val:long_name = "_FillValue example" ;
      fll_val:_FillValue = -999.f ;

  data:
    fll_val = 73, _, 73, _ ; 

} // group /
zender@firn:~$ 

```

**Comment by WardF** [comment]
Thanks @czender I recall there is another regression related to netcdf that is in `make test`; since no error code is returned our automated testing doesn't report an error, but I will take a look at both of those, starting with this one.

**Comment by DennisHeimbigner** [comment]
I have a question about the type(fillvalue) != type(variable) for packed variables.
I hypothesize two kinds of fillvalue.
1. define a fillvalue for the packed variable (of type e.g. short)  in defined.
    so type(fillvalue) == short
2. Interpret that the packed variable is actually representing an e.g. float and
    specifying type(fillvalue) == float
There is an obvious relationship if the float fillvalue is compressed to a packed value
and that packed value is used as the true fillvalue of the packed variable.

Judging from the example above, you are trying to do #2.
 Is that correct?

**Comment by DennisHeimbigner** [comment]
In thinking about all this, I realize I am working from the following
assumption: The netcdf-c library is not expected to have any knowledge about the
semantics of packing. That is, from the library's point of view,
it sees the following:
           short fll_val(lon) ;
               fll_val:add_offset = 73.f ;
That is, a variable of type short with a float valued attribute.
the name "add_offset" has no meaning to the library.

With respect to _FillValue, the semantics of _FillValue are known
to the library and have a specific semantics built into the library.
So, from the point of view of the library, this is wrong
               fll_val:_FillValue = -999.f ;
because of the type mismatch (float vs short).
This would be ok, though:
               fll_val:_FillValue = .XXXXs;
where XXXX is some short integer constant.

Given that the signature for
   nc_def_var_fill(int ncid, int varid, int no_fill, const void *fill_value)
has no provision for specifying the type of the fillvalue, it must of
necessity be using the type of the variable to interpret the bytes of
the fill_value argument.

So, I am baffled how specifying a fillvalue with type(fillvalue) != type(packed var)
could ever be implemented? What am I missing?

**Comment by czender** [comment]
Um, not sure exactly what you want to know. It may be easier to speak by phone. 949 891 2429. NCO never uses nc_def_var_fill(), as I was unaware such a function existed :) 
Remember, ncpdq was written in 2004, and with "missing_value" in mind, and the library had no semantic knowledge of "missing_value", though missing_value was far more common than _FillValue at the time. When packing a float into a short, ncpdq first defines the variable as a short, then _copies_ (with nc_copy_att()) all the metadata from the variable in in.nc to out.nc. Then it packs the floats into shorts, does nc_put_var() with the short data, then adds the scale_factor and add_offset attributes. ncpdq does nothing special with the _FillValue attribute. ncpdq never tries to pack or unpack attribute values. It just copies the _FillValue attribute and leaves it alone. Similarly, ncpdq does not try to pack or unpack _data_ that equals the _FillValue. Instead it does an implicity type conversion of the data from float to short. Hence if the _FillValue is -999.0f, then float data equal to the _FillValue are stored as -999s, because that is what a direct type conversion gives. When upacking, the reverse yields the right answer as long as the _FillValue is exactly representable as a short. Does this answer your question? If not please call...

**Comment by DennisHeimbigner** [comment]
No phone nearby.  Looks like I need to investigate how _FillValue is handled in the library
when doing copy etc. It should be case that one should not be able to define an attribute
named _FillValue at all because it has a leading underscore. This is beginning to look like
a problem of major proportions.

**Comment by WardF** [comment]
@DennisHeimbigner The documentation here: http://www.unidata.ucar.edu/software/netcdf/docs/fill_values.html says that _FillValue can be created directly, so I don't think it's a mistake.  It was never my understanding that you *couldn't* create an attribute with a leading underscore, but rather that that was used to denote a special attribute.  The page also says that we need to use a fill value of the same data type as the variable, but we will need to enforce that slowly and in concert with NCO.

**Comment by DennisHeimbigner** [comment]
You are correct, Ward. I did some further digging into the code.
What a mass of contradictions is netcdf. 
It turns out nc_def_var_fill is netcdf-4 only, it fails for a netcdf-3 file.
No obvious rationale, no idea why. Completely contradictory IMO.
So apparently, for netdf-3, the only way to set the fillvalue is to set the _FillValue attribute.
For netcdf-4, nc_def_var_fill overrides and replaces the _FillValue attribute; why have nc_def_var_fill
at all?
In any case, what ever we do for netcdf-3, we have to think about doing the same for
netcdf-4.
What a mess.

**Comment by WardF** [comment]
Dennis,

For the time being we'll back out to a known working state for the 4.5.0 release, and we will map out how to proceed at our next meeting.  Part of me wants to assume there is a good reason for this divergent behavior.  Personally I think we should continue with what netcdf3 does, i.e. allow for _FillValue to be set directly.  Of course, at this point, we have `nc_def_var_fill()` as well and we don't want to remove that, so we should probably add it for netcdf3 as well, although I'm happy to leave that for the 4.5.1 release.  Currently we have a lot of threads in the 4.5.0 release and I really want to get that out so I can start in on the backlog of pull requests and other issues.

**Comment by WardF** [comment]
I've created a docker image to help debug this issue.  When invoking docker it takes a local netcdf-c git folder as input and then runs the test script provided by @czender above.  

## Building/Using

This docker image depends on the `unidata/nctests:base` image.  If you do not have it locally, that is not a problem. Docker will download it automatically.

To build the docker image from the same directory containing these files:

    $ docker build -t unidata/nctests:nco -f Dockerfile.nco .

To run the image


    $ docker run --rm -it -v [path to netcdf-c directory]:/netcdf-c unidata/nctests:nco

    



[nco-docker.zip](https://github.com/Unidata/netcdf-c/files/1232556/nco-docker.zip)

**Comment by WardF** [comment]
The nice thing with this docker container is it will make running `git bisect` pretty straightforward, as well as just a general, sandboxed environment for testing incremental changes in libnetcdf against NCO.

**Comment by WardF** [comment]
# Git Bisect

## Running

Running git bisect as follows:

    $ git bisect start HEAD v4.4.1.1 --
    $ git bisect run docker run --rm -it -v $(pwd):/netcdf-c unidata/nctests:nco

## Results

Yields the following:

> 7fbde1b5bcdd3c41cab62b8ea3e4b1de6f40a7bc is the first bad commit
> commit 7fbde1b5bcdd3c41cab62b8ea3e4b1de6f40a7bc
> Author: Wei-keng Liao <wkliao@eecs.northwestern.edu>
> Date:   Tue Oct 25 15:09:52 2016 -0500
>
> add a new option ERANGE_FILL to use fill values when NC_ERANGE occurs; sync ncx.m4 with PnetCDF

## Conclusion

This suggests that the issue is related to one of the pull requests, as suspected.   I will investigate further *and* confirm that `git bisect` is correct.  There is a slim chance that the result was fouled due to a commit which didn't compile properly, but we will find out for sure and correct it.

**Comment by DennisHeimbigner** [comment]
I was just looking at the ERANGE_FILL pr. I did not realize
that he did it at the macro level.
Not good.
If we remove it, does that have any consequences for pnetcdf?

**Comment by DennisHeimbigner** [comment]
In any case, if the NCO code is not activating an ERANGE error, then
the ERANGE_FILL code should have no effect.

**Comment by WardF** [comment]
It appears this was misleading, as a later commit caacf4d3 is working.  I need to re-run the git bisect from this later commit.

**Comment by WardF** [comment]
Tagging in @wkliao in case he has any insight into this.

**Comment by wkliao** [comment]
I probably have more questions than "insights" :(

Regarding the global _FillValue attribute, I wonder how it is used in NCO or other NetCDF applications. According to the NetCDF documents, this attribute is intended to fill missing values of a variable. Without associated with a variable, what is the intent of having a global _FillValue?

Regarding the type matching of the variable and its _FillValue attribute, the NetCDF user guide describes the [Attribute Conventions](http://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html) and indicates that the type must be matched, even for packed variables. Quoted: "If the variable is packed using scale_factor and add_offset attributes, the _FillValue attribute should have the data type of the packed data."

Regarding to ERANGE_FILL issue, I guess this is most likely caused by the mismatched data types of a variable and its _FillValue attribute. Are the error messages available somewhere?

**Comment by czender** [comment]
@wkliao please comment on #463 if you know the answer.

Regarding your first question: One purpose of putting a _FillValue at the group level might be to indicate that is the intended _FillValue for all variables in the group.  I don't do or recommend this.

Regarding your comment: the attribute conventions did not, IIRC, say this in 2004 when I wrote ncpdq. And the immediate issue is backwards compatibility, not whether I was lazy in 2004. I was.

My application, ncpdq, does not throw ERANGE_FILL,  it throws  NC_EBADTYPE (-45): "NetCDF: Not a valid data type or _FillValue type mismatch" . Instructions to reproduce are in this thread.

**Comment by wkliao** [comment]
Hi, @czender
 
I wrote a test program below to check if NC_EBADTYPE is thrown when the data types are mismatched. As you can see, a variable is defined of type NC_FLOAT and its _FillValue attribute is NC_SHORT. When I compiled and ran against NetCDF 4.3.2, NC_EBADTYPE was returned at nc_enddef and nc_close. I am seeing from the source codes that nc_enddef() calls fill_NC_var() which checks the data types of variable and attribute _FillValue and returns NC_EBADTYPE when mismatched. I also found the same checking has even appeared in version 3.3 released in 1997. Could you give it a try?

```
#include <stdio.h>
#include <netcdf.h>

#define ERR {if(err!=NC_NOERR){printf("Error at line %d in %s: %s\n", __LINE__,__FILE__, nc_strerror(err));nerrs++;}}

int main(int argc, char *argv[])
{
    int err, nerrs=0, ncid, varid;
    short fillv=0;

    err = nc_create("test.nc", NC_CLOBBER, &ncid); ERR

    err = nc_def_var(ncid, "var", NC_FLOAT, 0, NULL, &varid); ERR

    err = nc_put_att_short(ncid, varid, _FillValue, NC_SHORT, 1, &fillv); ERR

    err = nc_enddef(ncid); ERR

    err = nc_close(ncid); ERR

    return (nerrs > 0);
}
```

**Comment by czender** [comment]
@wkliao I'm sure your program will give me the same result. I don't see the point of this particular exercise. As explained above, NCO uses a different code path (i.e., nc_copy_att()) than your program tests. Perhaps if you write a test with that codepath we can find which exact commit causes netCDF 4.5.x to change its behavior. We are trying to mitigate the shock to users if/when they link NCO to netCDF 4.5.0 and their workflows that involve packing with ncpdq suddenly break.

**Comment by wkliao** [comment]
Below is the code path that uses nc_copy_att.
Is this what NCO does?
```
#include <stdio.h>
#include <netcdf.h>

#define ERR {if(err!=NC_NOERR){printf("Error at line %d in %s: %s\n", __LINE__,__FILE__, nc_strerror(err));nerrs++;}}

int main(int argc, char *argv[])
{
    int err, nerrs=0, in_ncid, out_ncid, in_varid, out_varid;
    float fillv=0;

    err = nc_create("in.nc", NC_CLOBBER, &in_ncid); ERR
    err = nc_def_var(in_ncid, "var", NC_FLOAT, 0, NULL, &in_varid); ERR
    err = nc_put_att_float(in_ncid, in_varid, _FillValue, NC_FLOAT, 1, &fillv); ERR
    err = nc_close(in_ncid); ERR

    err = nc_open("in.nc", NC_NOWRITE, &in_ncid); ERR
    err = nc_inq_varid(in_ncid, "var", &in_varid); ERR

    err = nc_create("out.nc", NC_CLOBBER, &out_ncid); ERR
    err = nc_def_var(out_ncid, "var_out", NC_SHORT, 0, NULL, &out_varid); ERR
    err = nc_copy_att(in_ncid, in_varid, _FillValue, out_ncid, out_varid); ERR
    err = nc_enddef(out_ncid); ERR
    err = nc_close(out_ncid); ERR
    err = nc_close(in_ncid); ERR
    return (nerrs > 0);
}
```

**Comment by czender** [comment]
Yes, I think that captures the essence of what ncpdq does. But ncpdq throws the error in an nc_put_var() call that writes the (packed) data after nc_copy_att() and nc_enddef().

**Comment by wkliao** [comment]
When I ran that test program using v.4.3.2, I got NC_EBADTYPE at nc_enddef.
If you nc_set_fill() to set the fill mode to NC_NOFILL before nc_enddef, then there will be no error returned at nc_enddef. Can you check whether ncpdq disables fill mode?

**Comment by wkliao** [comment]
I found [line 763 of ncpdq.c](https://github.com/nco/nco/blob/master/src/nco/ncpdq.c#L763) turns off the fill mode.
```
763   nco_set_fill(out_id,NC_NOFILL,&fll_md_old);
```
I assume if this line is removed, ncpdq will fail when compiled against earlier NetCDF versions ... ?

**Comment by wkliao** [comment]
Just created a pull request #464 
It may silence the NC_EBADTYPE error for ncpdq.

**Comment by czender** [comment]
Yes, apparently the NC_NOFILL on line 763 prevents ncpdq packing mismatched _FillValue from failing with netCDF < 4.5.x. When I remove that line it fails with NC_EBADTYPE. Thanks for digging-in to this. I have not tested your patch, though it seems promising.

**Comment by WardF** [comment]
Thank you very much @wkliao for providing your insight and the patch; thank you also @czender for working with us to figure this out, and reporting it in the first place!  I've just tested #464 with the provided test, and so long as ERANGE_FILL is not enabled (default behavior), the test passes.  @czender, will this work for you? I suppose there is a possibility of issues being reported if somebody goes out of their way to turn on ERANGE_FILL, but since it's not the default behavior I feel comfortable calling this fixed.  I've pushed a new branch, `wkliao-ghpull464`; once the CI tests finish, I will merge it back into `v4.5.0-release-branch` and from there it will propagate back into `master`.

**Comment by czender** [comment]
Please ping me when @wkliao 's patch is in master and I'll test. Assuming it passes, then this is a fine solution for ncpdq users.  Thank you Wei-Keng, for producing this so quickly.

**Comment by WardF** [comment]
@czender This is in `master`, let me know if it works for you and if so I'll close this out and move on to the other cdf5-related issues.

**Comment by czender** [comment]
This thread discusses two distinct changes in _FillValue behavior in 4.5.x, each of which broke one NCO regression test that passed with 4.4.x and earlier libraries. Today's master still causes an error when attempting to create a _FillValue for NC_GLOBAL. I mentioned that I patched NCO to continue (not abort) despite this error:
```
zender@skyglow:~/nco$ ncatted -O -a _FillValue,global,c,l,222 ~/nco/data/in.nc ~/foo.nc
WARNING: nco_put_att() received NC_EGLOBAL error writing attribute "_FillValue" to metadata for group "/". netCDF 4.5.x forbids this with the _FillValue attribute, though earlier versions allow it. Proceeding normally without writing _FillValue attribute...
```
Personally I can live with the current behavior because I don't use/need this feature. As discussed above, Unidata needs to decide whether to allow this feature to work become some users may rely on it, and the new behavior would reduce backwards compatibility for them.

As for the second issue, which was diagnosed as causing ncpdq to fail, the current snapshot does fix that issue (thanks to @wkliao for the patch). This will prevent massive breakage in many workflows so this is an important patch to keep. In the future I would like to alter ncpdq so it does not rely on this patch, but that would be a big project that I don't have time for right now.

**Comment by WardF** [comment]
I think we are going to allow it; it has been allowed this long and has become permissible, so I think the documentation needs to be updated.  I thought I had reverted this behavior already but apparently not.  I will look into it.  

Actually, yes I've reverted it in a branch but have not met with @DennisHeimbigner yet to discuss/confirm.  And yes, thank you @wkliao for the patch with ncpdq.

@czender - NCO is super important, and breaking it is not something I want to do.  I've built a pretty robust automated regression/continuous integration framework based on docker, but the weakness is that it's automated; if a test fails and returns 0, it will not catch it.  I'd like to figure out a way to catch some of these netcdf tests that only run via `make test` and not `make check`.  Do you have any ideas? Worst case scenario would be to write a string parser for the test output and use that, but it's unlikely I'd get to that in the near future.  Could some of these netcdf specific tests be moved into 'make check', maybe? Or would that be a departure from your development design?

**Comment by wkliao** [comment]
Please note that as early as version 3.3 released in 1997, checking type mismatch has already appeared in libsrc/putget.m4. This checking happens during nc_enddef and when fill mode is on.

```
 126         /*
 127          * Set up fill value
 128          */
 129         attrpp = NC_findattr(&varp->attrs, _FillValue);
 130         if( attrpp != NULL )
 131         {
 132                 /* User defined fill value */
 133                 if( (*attrpp)->type != varp->type || (*attrpp)->nelems != 1 )
 134                 {
 135                         return NC_EBADTYPE;
 136                 }
```

**Comment by DennisHeimbigner** [comment]
I agree that we should continue to allow _FillValue with NC_GLOBAL.
We just need to note that it is (for now) ignored by the netcdf library.

**Comment by wkliao** [comment]
FYI. I found the following statement from netcdf man page of v 3.3, libsrc/netcdf.3

```
798 By default, the netCDF interface sets the values of
799 all newly-defined variables of finite length (i.e. those that do not have
800 an unlimited, dimension) to the type-dependent fill-value associated with each
801 variable.  This is done when \fBnc_endef(\|)\fR
802 is called.  The
803 fill-value for a variable may be changed from the default value by
804 defining the attribute `\fB_FillValue\fR' for the variable.  This
805 attribute must have the same type as the variable and be of length one.
```

**Comment by czender** [comment]
FYI, this command fails with 4.5.0-rc3 and succeeds in 4.4.x. I thought Unidata intended to continue to let users write _FillValue as group metadata in 4.5.0. The current release candidate does not allow that.

```
zender@skyglow:~$ ncks --lbr
Linked to netCDF library version 4.5.1-development compiled Oct  1 2017 16:21:13
zender@skyglow:~$ ncatted -O -a _FillValue,global,c,l,222 ~/nco/data/in.nc ~/foo.nc
WARNING: nco_put_att() received NC_EGLOBAL error writing attribute "_FillValue" to metadata for group "/". netCDF 4.5.x forbids this with the _FillValue attribute, though earlier versions allow it. Proceeding normally without writing _FillValue attribute...
```

```
zender@aerosol:~$ ncks --lbr
Linked to netCDF library version 4.4.1.1 compiled Jul 15 2017 09:29:02
zender@aerosol:~$ ncatted -O -a _FillValue,global,c,l,222 ~/nco/data/in.nc ~/foo.nc
zender@aerosol:~$ 
```

**Comment by WardF** [comment]
That as the intent, and I thought I had reversed the pull request disallowing this.  I even thought I had added a test for this. Let me go double-check.  This should be allowed.

**Comment by DennisHeimbigner** [comment]
This may also require a change to ncgen; I will investigate separately.

**Comment by czender** [comment]
Please re-read my message of August 24. Nothing seems to have changed since then.

**Comment by WardF** [comment]
Thanks @czender we will get this in!

**Comment by WardF** [comment]
Ok, looking at this we had corrected the initial behavior using an ncpdq-based test.  Looking at this ncatted issue to see what's going on.

**Comment by DennisHeimbigner** [comment]
It looks like there is a check in libdispatch/dattput.c
in (at least) each of the nc_put_att_XXX routines,
where XXX is some type (e.g. int or float or...).
You should be able to find all of them by looking for
occurrences of NC_EGLOBAL.

**Comment by WardF** [comment]
I've got this taken care of and am adjusting a test.  I'll roll this in soon today and get ball rolling on the 4.5.0 full release.

**Comment by WardF** [comment]
The fix is merged into v4.5.0-release-branch. Once travis-ci completes, I
will merge into master.

**Comment by WardF** [comment]
I believe we have this addressed for 4.5.0.  Tagging to revisit for 4.5.1.

**Comment by WardF** [comment]
It appears this is fixed, closing out as part of issue cleanup, will reopen if incorrect.
